src/v/raft/seq_stm.cc

```diff
@@ -0,0 +1,337 @@
+              .then([this]() { _is_catching_up = false; });
+        }
+    }
+    return ss::now();
+}
+
+static int64_t time_since_epoch() {
+    return std::chrono::duration_cast<std::chrono::milliseconds>(
+             clock_type::now().time_since_epoch())
+      .count();
+}
+
+ss::future<checked<raft::replicate_result, kafka::error_code>>
+seq_stm::replicate(
+  raft::batch_identity bid,
+  model::record_batch_reader&& r,
+  raft::replicate_options opts) {
+    if (bid.pid.id >= 0) {
+        if (_insync_term != _c->term()) {
+            (void)catchup();
+            return seastar::make_ready_future<
+              checked<replicate_result, kafka::error_code>>(
+              kafka::error_code::not_leader_for_partition);
+        } else {
+            auto pid_seq = _seq_table.find(bid.pid);
+            auto last_write_timestamp = time_since_epoch();
+            if (pid_seq == _seq_table.end()) {
+                if (bid.seq == 0) {
+                    seq_entry entry{
+                      .pid = bid.pid,
+                      .seq = bid.record_count - 1,
+                      .last_write_timestamp = last_write_timestamp};
+                    _oldest_session = std::min(
+                      _oldest_session,
+                      model::timestamp(entry.last_write_timestamp));
+                    _seq_table.emplace(entry.pid, entry);
+                } else {
+                    return seastar::make_ready_future<
+                      checked<replicate_result, kafka::error_code>>(
+                      kafka::error_code::out_of_order_sequence_number);
+                }
+            } else {
+                if (pid_seq->second.seq + 1 == bid.seq) {
+                    pid_seq->second.seq += bid.record_count;
+                    pid_seq->second.last_write_timestamp = last_write_timestamp;
+                    _oldest_session = std::min(
+                      _oldest_session,
+                      model::timestamp(pid_seq->second.last_write_timestamp));
+                } else {
+                    return seastar::make_ready_future<
+                      checked<replicate_result, kafka::error_code>>(
+                      kafka::error_code::out_of_order_sequence_number);
+                }
+            }
+            return _c->replicate(_insync_term, std::move(r), opts)
+              .then([](result<replicate_result> r) {
+                  if (r) {
+                      return checked<replicate_result, kafka::error_code>(
+                        r.value());
+                  } else {
+                      return checked<replicate_result, kafka::error_code>(
+                        kafka::error_code::unknown_server_error);
+                  }
+              });
+        }
+    }
+    return _c->replicate(std::move(r), std::move(opts))
+      .then([](result<replicate_result> r) {
+          if (r) {
+              return checked<replicate_result, kafka::error_code>(r.value());
+          } else {
```

Comment:
no need for else.

Reply:
there are few other places where `else` is not needed.